espnet2/speechlm/core_lm/ar_parallel.py

```diff
@@ -0,0 +1,109 @@
+
+from typing import Dict, Tuple
+
+import torch
+
+from espnet2.speechlm.core_lm.abs_core_lm import AbsCoreLM
+from espnet2.speechlm.inference_utils import AbsInferenceConfig
+
+
+class ARParallelLM(AbsCoreLM):
+    def __init__(
+        self,
+        transformer,
+        continuous_encoders: dict,
+        vocab_size: int,
+        nq: int,
+        share_emb: bool = False,
+    ):
+        """Initialize Auto-regressive LM with parallel interleave codec pattern.
+
+        Args:
+            transformer (torch.nn.Module): the Transformer body implementation
+            vocab_size (int): Dimention of vocabulary.
+            nq (int): Number of codes for each token / frame, usually for speech codec.
+            share_emb (bool): If true, share the embedding and lm_head weight.
+        """
+        super(ARParallelLM, self).__init__()
+
+        self.decoders = transformer
+
+        self.emb = torch.nn.Embedding(vocab_size, transformer.d_model, padding_idx=0)
+
+        self.lm_head = torch.nn.Linear(transformer.d_model, vocab_size, bias=False)
+        if share_emb:
+            self.lm_head.weight = self.emb.weight
+
+        # TODO(Jinchuan): force self.head_emb.weight[0] = 0
+        self.head_emb = torch.nn.Embedding(12, transformer.d_model, padding_idx=0)
+
+        if hasattr(self.decoders, "init_embeddings"):
+            self.decoders.init_embeddings(self.emb, self.lm_head)
+
+        self.continuous_encoders = torch.nn.ModuleDict(continuous_encoders)
+        for _, module in continuous_encoders.items():
+            module.register_projection(odim=transformer.d_model)
+
+        self.nq = nq
+        self.n_ctx = transformer.n_ctx
+
+    def forward(
+        self,
+        dec_seq: torch.Tensor,
+        loss_mask: torch.Tensor = None,
+        conti_feats: list = None,
+    ) -> Tuple[torch.Tensor, Dict, torch.Tensor]:
+        """Auto-Regresive LM forward for training
+
+        Args:
+            dec_seq (LongTensor): Batch of decoder sequences (B, T, nq).
+            loss_mask (LongTensor): Lengths of condition part in dec_seq (B, T, nq).
+        """
+        assert dec_seq.dim() == 3
+
+        targets = dec_seq[:, 1:]
+        loss_mask = loss_mask[:, 1:] if loss_mask is not None else None
+        x = dec_seq[:, :-1]
+
+        # input embedding
+        x = self.emb(x).sum(dim=2)
+        if len(self.continuous_encoders) > 0:
+            for modality, module in self.continuous_encoders.items():
+                x = module(
+                    x,
+                    conti_feats,
+                    modality=modality,
+                )
+
+        # transformer output
+        x = self.decoders(x)
+        x = x.unsqueeze(2) + self.head_emb.weight.tile(1, 1, 1, 1)[:, :, : self.nq]
+
+        # NOTE(Jinchuan): We don't apply lm_head here naively. It is implemented in
+        # loss module to save computing.
+        return x, targets, loss_mask
+
+    @torch.no_grad()
+    def inference(
```

Comment:
Will you add this in the future?
If so, it's better to add TODO

Reply:
Add comment:

TODO(Jinchuan):
        We may not implement this inference code in the near future, as the parallel
        implementation would give poor performance. Anyone who is interested in 
        parallel should consider using ar_delay.py instead.

        Potentially, the parallel would be better aligned with the FSQ codec. Users
        who are interested in this direction could consider implementing the inference
        by themselves.